### PR TITLE
feat(sequencer): store sequencer blocks in the sequencer state

### LINKED ISSUE #773: sequencer: generate and store block structure and proofs in sequencer app for usage in API
the sequencer app should store the `SequencerBlock` structure and related merkle tree proofs in the app storage, so that it doesn't need to be recalculated each time the sequencer API is called.

### PR DESCRIPTION
**Pull request by noot**
## Summary
generate and store `SequencerBlock`s in the sequencer app state for usage in the sequencer API. now, the sequencer API does not need to call cometbft and do heavy conversions, it just reads the `SequencerBlock`s stored during app execution from state.

## Background
see above - less conversions and eventually we can remove the need for relayer/conductor to call cometbft at all.

## Changes
- implement the `api_state_ext` module which writes a sequencer block to state by splitting it into its different parts (header, rollup datas, proofs) and writing those to disk separately. when fetching a filtered sequencer block, only the relevant data needs to be fetched.
- allows fetching a block/data by block number or hash.
- implement the `SequencerBlockHeader` type which contains the cometbft header and the relevant rollup data commitments, which really *should* be part of the block header but we haven't forked cometbft. these fields need to exist in the `SequencerBlock` type as they need to written to disk, and the commitments need for the filtered sequencer block API. if they're not stored, then fetching filtered block data will require the entire block data to be fetched so that the commitments can be reconstructed, which is not what we want.
- inside the app, implement `SequencerBlockBuilder` which stores the header during `begin_block` and each tx during `deliver_tx`, and in `end_block` uses these to construct and store the `SequencerBlock`. note: this can be removed after #679 is implemented

## Testing
unit tests + i ran it with cometbft. (TODO: I want to add more unit tests to the sequencer API itself)

## Related Issues

closes #773

### CODE REVIEW

File: crates/astria-sequencer/src/grpc/sequencer.rs
Comment (joroshiba) [inline crates/astria-sequencer/src/grpc/sequencer.rs:146]: Does `FilteredSequencerBlock` need `all_rollup_ids` or does it only need the rollup_ids root + proof?

**Review by joroshiba — COMMENTED** [review]
Couple nits overall looks good, only two small things that need to be addressed here. 

Higher level question on introduction of the new header type and where that should propogate to:

`FilteredSequencerBlock` and `SequencerBlock` are currently out of sync in terms of how the handle adding this information. It feels like they should match?

I'm good with the idea of a `SequencerBlockHeader` containing the roots, it makes logical sense, and works well for the state storage, but should we be more consistent with it?

Additionally worth thinking about this all the way down. We have the `CelestiaSequencerBlob` and `CelestiaRollupBlob` types as well. What is the translation pipeline, which of these should be similar in shape? If we stick with something more flat at the API level, ie not a custom type here for the header it matches what we are doing elsewhere more and isn't a breaking proto change. That doesn't mean it's not the right overall decision to have the new shape.

File: crates/astria-sequencer/src/api_state_ext.rs
Comment (noot) [reply crates/astria-sequencer/src/api_state_ext.rs:36]: yes, will fix

File: crates/astria-sequencer/src/grpc/sequencer.rs
Comment (noot) [reply crates/astria-sequencer/src/grpc/sequencer.rs:61]: it could be either not found or unable to read the state so I left as internal (more general)

File: crates/astria-sequencer/src/grpc/sequencer.rs
Comment (noot) [reply crates/astria-sequencer/src/grpc/sequencer.rs:146]: it'll still need all the rollup IDs for conductor, so conductor can verify if its rollup ID was included or not (and not omitted in the case of an untrusted/malicious sequencer node)

**Review by joroshiba — APPROVED** [review]

**Review by SuperFluffy — CHANGES_REQUESTED** [review]
Instead of review comments I made #793. Please have a look + merge, will approve after.

**Review by SuperFluffy — APPROVED** [review]
approve after merging the fixup PR

**Pull request merged**
